File: my_py_actions/my_py_actions/count_until_server.py
#!/usr/bin/env python3

from typing import List
import time
import threading
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from rclpy.action.server import ServerGoalHandle, GoalResponse, CancelResponse
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup

from my_robot_interfaces.action import CountUntil


class CountUntilServerNode(Node):
    def __init__(self):
        super().__init__("count_until_server")

        self.goal_handle_ = None
        self.goal_lock_ = threading.Lock()

        self.goal_queue_: List[ServerGoalHandle] = []

        self.count_until_server_ = ActionServer(
            node=self,
            action_type=CountUntil,
            action_name="count_until",
            goal_callback=self.goal_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            execute_callback=self.execute_callback,
            cancel_callback=self.cancel_callback,
            callback_group=ReentrantCallbackGroup(),
        )

        self.get_logger().info("Count Until Server started.")

    def goal_callback(self, goal: CountUntil.Goal):
        self.get_logger().info("Received a goal.")

        # # Policy: refuse new goal if another is active
        # with self.goal_lock_:
        #     if self.goal_handle_ is not None and self.goal_handle_.is_active:
        #         self.get_logger().info(
        #             "Rejecting goal because a goal is already active."
        #         )
        #         return GoalResponse.REJECT

        if goal.target <= 0:
            self.get_logger().info("Rejecting goal because target is not positive.")
            return GoalResponse.REJECT

        if goal.period <= 0:
            self.get_logger().info("Rejecting goal because period is not positive.")
            return GoalResponse.REJECT

        # # Policy: preempt previous goal when receiving new goal
        # with self.goal_lock_:
        #     if self.goal_handle_ is not None and self.goal_handle_.is_active:
        #         self.get_logger().info("Aborting previous goal.")
        #         self.goal_handle_.abort()

        self.get_logger().info("Accepting goal.")
        return GoalResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            if self.goal_handle_ is not None:
                self.get_logger().info("Queuing goal.")
                self.goal_queue_.append(goal_handle)
            else:
                goal_handle.execute()

    def cancel_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Received a cancel request.")
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle

        request: CountUntil.Goal = goal_handle.request
        target = request.target
        period = request.period

        self.get_logger().info(f"Counting until {target} with period {period}.")

        result = CountUntil.Result()
        result.count = 0

        feedback = CountUntil.Feedback()
        feedback.count = result.count
        goal_handle.publish_feedback(feedback)

        for i in range(target):
            time.sleep(period)

            if not goal_handle.is_active:
                self.execute_next_goal()
                return result

            if goal_handle.is_cancel_requested:
                self.get_logger().info("Cancelling goal.")
                goal_handle.canceled()
                self.execute_next_goal()
                return result

            result.count += 1
            self.get_logger().info(f"Counted to {result.count}.")

            feedback.count = result.count
            goal_handle.publish_feedback(feedback)

        goal_handle.succeed()
        self.execute_next_goal()
        return result

    def execute_next_goal(self):
        with self.goal_lock_:
            if len(self.goal_queue_) > 0:
                self.get_logger().info("Executing next goal.")
                self.goal_queue_.pop(0).execute()
            else:
                self.goal_handle_ = None


def main(args=None):
    rclpy.init(args=args)

    node = CountUntilServerNode()
    rclpy.spin(node, MultiThreadedExecutor())

    rclpy.shutdown()


if __name__ == "main":
    main()

File: my_cpp_actions/src/count_until_server.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>

#include <my_robot_interfaces/action/count_until.hpp>

using namespace std::placeholders;

class CountUntilServerNode : public rclcpp::Node
{
public:
  CountUntilServerNode() : Node("count_until_server")
  {
    pCallbackGroup_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);

    pActionServer_ =
        rclcpp_action::create_server<my_robot_interfaces::action::CountUntil>(
            this, "count_until",
            std::bind(&CountUntilServerNode::cbGoal, this, _1, _2),
            std::bind(&CountUntilServerNode::cbCancel, this, _1),
            std::bind(&CountUntilServerNode::cbHandleAccepted, this, _1),
            rcl_action_server_get_default_options(),
            pCallbackGroup_);

    RCLCPP_INFO(this->get_logger(), "Count Until Server started.");
  }

private:
  rclcpp_action::GoalResponse cbGoal(const rclcpp_action::GoalUUID & /*uuid*/, my_robot_interfaces::action::CountUntil::Goal::ConstSharedPtr pGoal)
  {
    RCLCPP_INFO(this->get_logger(), "Received a goal.");

    //// Policy: refuse new goal if another is active
    // {
    //   std::lock_guard<std::mutex> lock(mutexGoalHandle_);
    //   if (pGoalHandle_ && pGoalHandle_->is_active())
    //   {
    //     RCLCPP_INFO(this->get_logger(), "Rejecting goal because a goal is already active.");
    //     return rclcpp_action::GoalResponse::REJECT;
    //   }
    // }

    if (pGoal->target <= 0)
    {
      RCLCPP_INFO(this->get_logger(), "Rejecting goal because target is not positive.");
      return rclcpp_action::GoalResponse::REJECT;
    }

    if (pGoal->period <= 0)
    {
      RCLCPP_INFO(this->get_logger(), "Rejecting goal because period is not positive.");
      return rclcpp_action::GoalResponse::REJECT;
    }

    //// Policy: preempt previous goal when receiving new goal
    // {
    //   std::lock_guard<std::mutex> lock(mutexGoalHandle_);
    //   if (pGoalHandle_ && pGoalHandle_->is_active())
    //   {
    //     RCLCPP_INFO(this->get_logger(), "Aborting previous goal.");
    //     preemptingGoalID_ = pGoalHandle_->get_goal_id();
    //   }
    // }

    RCLCPP_INFO(this->get_logger(), "Accepting goal.");
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  rclcpp_action::CancelResponse cbCancel(const std::shared_ptr<rclcpp_action::ServerGoalHandle<my_robot_interfaces::action::CountUntil>> /*pGoalHandle*/)
  {
    RCLCPP_INFO(this->get_logger(), "Received a cancel request.");
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void cbHandleAccepted(const std::shared_ptr<rclcpp_action::ServerGoalHandle<my_robot_interfaces::action::CountUntil>> pGoalHandle)
  {
    {
      std::lock_guard<std::mutex> lock(mutexGoalHandle_);
      pGoalHandle_ = pGoalHandle;
    }
    executeGoal(pGoalHandle);
  }

  void executeGoal(const std::shared_ptr<rclcpp_action::ServerGoalHandle<my_robot_interfaces::action::CountUntil>> pGoalHandle)
  {
    const auto target = pGoalHandle->get_goal()->target;
    const auto period = pGoalHandle->get_goal()->period;

    auto pResult = std::make_shared<my_robot_interfaces::action::CountUntil::Result>();
    pResult->count = 0;

    auto pFeedback = std::make_shared<my_robot_interfaces::action::CountUntil::Feedback>();
    pFeedback->count = pResult->count;
    pGoalHandle->publish_feedback(pFeedback);

    rclcpp::Rate loopRate(1.0 / period);
    for (int i = 0; i < target; ++i)
    {
      loopRate.sleep();

      // {
      //   std::lock_guard<std::mutex> lock(mutexGoalHandle_);
      //   if (pGoalHandle->get_goal_id() == preemptingGoalID_)
      //   {
      //     pGoalHandle->abort(pResult);
      //     return;
      //   }
      // }

      if (pGoalHandle->is_canceling())
      {
        RCLCPP_INFO(this->get_logger(), "Cancelling goal.");
        pGoalHandle->canceled(pResult);
        return;
      }

      RCLCPP_INFO(this->get_logger(), "Counted to %ld.", ++(pResult->count));

      pFeedback->count = pResult->count;
      pGoalHandle->publish_feedback(pFeedback);
    }

    pGoalHandle->succeed(pResult);
  }

  rclcpp_action::Server<my_robot_interfaces::action::CountUntil>::SharedPtr pActionServer_;
  rclcpp::CallbackGroup::SharedPtr pCallbackGroup_;

  std::mutex mutexGoalHandle_;
  std::shared_ptr<rclcpp_action::ServerGoalHandle<my_robot_interfaces::action::CountUntil>> pGoalHandle_;

  // rclcpp_action::GoalUUID preemptingGoalID_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  auto pNode = std::make_shared<CountUntilServerNode>();

  rclcpp::executors::MultiThreadedExecutor executor;
  executor.add_node(pNode);
  executor.spin();

  // rclcpp::spin(std::make_shared<CountUntilServerNode>());

  rclcpp::shutdown();
  return 0;
}
